ipam/pool_test.go

```
@@ -0,0 +1,991 @@
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When as has one pool available", func() {
+				It("Should return the pool", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id: poolId,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal(poolId))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When as has pools with different priority", func() {
+				It("Should return the pool with the highest priority", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["pool1"] = &addressPool{
+						Id:       "pool1",
+						Priority: 1,
+					}
+					as.Pools["pool2"] = &addressPool{
+						Id:       "pool2",
+						Priority: 2,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("pool2"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When as has pools with different addresses", func() {
+				It("Should return the pool with the highest number of addresses", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["pool1"] = &addressPool{
+						Id:        "pool1",
+						Addresses: map[string]*addressRecord{},
+					}
+					as.Pools["pool2"] = &addressPool{
+						Id:        "pool2",
+						Addresses: map[string]*addressRecord{},
+					}
+					as.Pools["pool2"].Addresses["ar"] = &addressRecord{}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("pool2"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+		})
+
+		Describe("Test releasePool", func() {
+			Context("When pool not found", func() {
+				It("Should return error", func() {
```

```suggestion
				It("Should return an error", func() {
```

---

"Should raise an error?" seems better?